helsingborg-stad/app-mitt-helsingborg
# Feat/testa cicd pipeline med fastlane cu fx7a9v

## Explain the changes you’ve made

Added base setup for Fastlane so that production builds can be produced (and optionally uploaded) with minimal effort. This is also made in preparation for GitHub Actions integration.

In practice this adds configuration files required for Fastlane to build a production `ipa` file, as well as adds Gemfiles to handle ruby dependencies.

Currently this is iOS-only.

## Explain why these changes are made

These changes are part of an automation system for build releases which will help developers reduce time spent on deployment and release tasks. This also facilitates building a system that performs some automatic checks on pending pull requests.

## Explain your solution

Fastlane is used as automation framework, which takes care of the heavy lifting such as configuring the xcode workspace and running the correct commands. Configuration is made with environment variables in `/ios/fastlane/.env`.

## How to test the changes?

Only works on Mac. Requires the computer to be able to build the project manually (requires XCode, build tools etc.).

Ruby, Fastlane, and Cocoapods are required. It is recommended to match the versions used by the intended CI-system ([GitHub Actions macos-latest](https://github.com/actions/virtual-environments/blob/main/images/macos/macos-10.15-Readme.md)), which are:

* MacOS 10.15.7
* Xcode 12.4
* Xcode Command Line Tools 12.4.0.0.1.1610135815
* Ruby 2.7.3
* Rubygems 3.2.16
* Fastlane 2.180.1
* Cocoapods 1.10.1

It is recommended (but not required) to manage these dependencies with rbenv and bundler. To use:

__Install Ruby__

1. Run `brew install rbenv` - this will install rbenv.
2. From project root run `rbenv init` (note: this needs to be done for every new shell; follow the instructions from `rbenv init` to add to your shell startup script).
3. Run `rbenv install` - this will install the ruby version specified in `.ruby-version` in the current dir.
4. Verify version with `ruby --version`.

__Install Gems__

1. Run `gem install bundler` - this will install bundler.
2. Run `bundle install` - this will install the dependencies from `Gemfile`.
3. Verify versions with `fastlane --version` and `pod --version`.

__Test Fastlane__

1. Copy `/ios/fastlane/example.env` to `/ios/fastlane/.env` and fill in the blanks (instructions are in the file). For current shared settings ping @KenthLjungHBG  👋 .
2. From `/ios` run `fastlane build`, this will run the _build_ lane.
3. Follow the instructions. First time you have to enter some credentials, but consecutive runs should not require input (run the build lane again afterwards to test).
4. Run `fastlane build` again - this will re-build but should use any existing cache (`/ios/.local_derived_data`). 
5. If everything is working, you should see a `MittHelsingborg.ipa` file in `/ios` as well as a few other extra files.

## Was this feature tested in the following environments?

- [❌] Storybook on a iOS device/simulator.
- [✅] Building the Application on a iOS device/simulator.
- [❌] Building the Application on an Android device/simulator.

Tested with:
* MacOS Big Sur 11.2.3 (20D91)
* Xcode Version 12.4 (12D4e)
* ruby 2.7.3p183 (2021-04-05 revision 6847ee089d) [x86_64-darwin20]
* gem 3.1.6
* Bundler version 2.2.16
* fastlane 2.180.1

## Anything else? (optional)

There is also a beta lane available (`bundle exec fastlane beta`) which will build and upload a TestFlight build. **Note** however that it is **not recommended** to run this lane as it will (if successful) upload new permanent builds to App Store Connect. Builds are not active (published/notified) by default however they will clutter up the build section and affect build numbers. This lane is intended to be run on a CI system in the future.

Note also that the version string is not managed via Fastlane, so if uploading to TestFlight you may have to manually set the version string in XCode to be the same or greater than the latest upload.
